runtime_services/src/runtime_services.rs

```diff
@@ -0,0 +1,864 @@
+#![cfg_attr(all(not(test), not(feature = "mockall")), no_std)]
+
+extern crate alloc;
+
+pub mod variable_services;
+
+#[cfg(any(test, feature = "mockall"))]
+use mockall::automock;
+
+use alloc::vec::Vec;
+use core::{
+    ffi::c_void,
+    marker::PhantomData,
+    ptr,
+    sync::atomic::{AtomicPtr, Ordering},
+};
+
+use r_efi::efi;
+pub use variable_services::{GetVariableStatus, VariableIdentifier, VariableInfo, VariableNameIterator};
+
+/// This is the runtime services used in the UEFI.
+/// it wraps an atomic ptr to [`efi::RuntimeServices`]
+#[derive(Debug)]
+pub struct StandardRuntimeServices<'a> {
+    efi_runtime_services: AtomicPtr<efi::RuntimeServices>,
+    _lifetime_marker: PhantomData<&'a efi::RuntimeServices>,
+}
+
+impl<'a> StandardRuntimeServices<'a> {
+    /// Create a new StandardRuntimeServices with the provided [efi::RuntimeServices].
+    pub const fn new(efi_runtime_services: &'a efi::RuntimeServices) -> Self {
+        // The efi::RuntimeServices is only read, that is why we use a non mutable reference.
+        Self {
+            efi_runtime_services: AtomicPtr::new(efi_runtime_services as *const _ as *mut _),
+            _lifetime_marker: PhantomData,

```

Comment:
I thought @Javagedes had described a way where we could get away from PhantomData, but perhaps I am misremembering or it doesn’t apply in this case?

Reply:
Not the case here, PhantomData is being used here to make the rust compiler borrowing rules stricter and safer by saying `StandardRuntimeServices` must live as long as `efi::RuntimeServices`. Without this change, the efi_runtime_services could go out of scope and the `StandardRuntimeServices` object would never know because We just take an AtpomicPtr to the address it exists at and call it a day.

What this change does is two fold:

1. It allows us to catch the possibility of the pointer being NULL at compile time instead of runtime as the lifetimes "say so".
2. Because of this, we can guarantee that the pointer is never NULL, and thus can skip all the `is_null()` checks we would normally have to do.